lib/modules/coverage/contract_source.js

```diff
@@ -0,0 +1,306 @@
+const SourceMap = require('./source_map');
+
+class ContractSource {
+  constructor(file, body) {
+    this.file = file;
+    this.body = body;
+
+    this.lineLengths = body.split("\n").map((line) => { return line.length; });
+    this.lineCount = this.lineLengths.length;
+
+    this.lineOffsets = [];
+    this.lineLengths.forEach((length, line) => {
+      if(line == 0) {
+        this.lineOffsets[0] = 0;
+        return;
+      }
+
+      // +1 here factors in newline characters.
+      this.lineOffsets[line] = this.lineOffsets[line-1] + this.lineLengths[line-1] + 1;
+    });
+
+    this.contracts = {};
+  }
+
+  sourceMapToLocations(sourceMap) {
+    var [offset, length, ..._] = sourceMap.split(":").map((val) => {
+      return parseInt(val, 10);
+    });
+
+    var locations = {};
+
+    for(let i = 0; i < this.lineCount; i++) {
+      if(this.lineOffsets[i+1] <= offset) continue;
+
+      locations.start = {line: i, column: offset - this.lineOffsets[i]};
+      break;
+    }
+
+    for(var i = locations.start.line; i < this.lineCount; i++) {
```

Comment:
I did this originally but it was hard to make sense of what was going on. You'll notice that I offset the second loop so that we're not doing any unnecessary iterations anyway, so this is as optimized as it gets.